### PR TITLE
[expo-av][web] Fix Video Fullscreen APIs for Safari

### LINKED ISSUE #12257: [expo-av][web] Fullscreen APIs for Video broken in Safari on iOS and macOS
### Summary

The `Video.onFullscreenUpdate`,`videoRef.presentFullscreenPlayer`, and `videoRef.dismissFullscreenPlayer` APIs are broken in Safari on both iOS and macOS. The `videoRef.presentFullscreenPlayer` API throws the error `Unhandled Rejection (TypeError): element.requestFullscreen is not a function. (In 'element.requestFullscreen()', 'element.requestFullscreen' is undefined)`. The other APIs are difficult to test without that API working, but after a little experimentation I was able to determine that they also are broken (`onFullscreenUpdate` is never called even when a video is switched to fullscreen and `dismissFullscreenPlayer` throws an error similar to `presentFullscreenPlayer`).

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Web

### SDK Version (managed workflow only)

_No response_

### Environment

Expo CLI 4.3.0 environment info:
    System:
      OS: macOS 11.2.2
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 15.11.0 - ~/.nvm/versions/node/v15.11.0/bin/node
      Yarn: 1.22.10 - ~/.nvm/versions/node/v15.11.0/bin/yarn
      npm: 7.6.0 - ~/.nvm/versions/node/v15.11.0/bin/npm
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7042882
      Xcode: /undefined - /usr/bin/xcodebuild
    npmPackages:
      expo: ~41.0.0-alpha.0 => 41.0.0-alpha.0 
      react-native: 0.63.2 => 0.63.2 
    npmGlobalPackages:
      expo-cli: 4.3.0
    Expo Workflow: bare

### Reproducible demo or steps to reproduce from a blank project

1) Setup expo for local development by following the steps in CONTRIBUTIONS.md.
2) Run `yarn web` from `apps/bare-expo`.
3) Open the demo app in Safari (macOS or iOS).
4) Navigate to "Components" > "Video".
5) Click/tap "Open fullscreen" and observe the React error.

### PR DESCRIPTION
**Pull request by elliotdickison**
**Problem**

The `Video.onFullscreenUpdate`, `videoRef.presentFullscreenPlayer`, and `videoRef.dismissFullscreenPlayer`
APIs are broken in Safari on both iOS and macOS. The `videoRef.presentFullscreenPlayer` API throws the
error `Unhandled Rejection (TypeError): element.requestFullscreen is not a function. (In
'element.requestFullscreen()', 'element.requestFullscreen' is undefined)`. The other APIs are difficult
to test without that API working, but after a little experimentation I was able to determine that they
also are broken (`onFullscreenUpdate` is never called even when a video is switched to fullscreen with the working native control and `dismissFullscreenPlayer` throws an error similar to `presentFullscreenPlayer`).

**Solution**

I extracted the code that interacts with the browser's fullscreen APIs into a separate
WebFullscreenUtils.ts module (named similarly to other web-only utility modules like WebCameraUtils.ts and
WebCapabilityUtils.ts). I added some feature-detection code to these APIs to switch between the standard
fullscreen API (the previous solution) and the non-standard webkit fullscreen API (my addition to fix
fullscreen usage in Safari).

**Testing**

For each of the browsers listed below I manually:
1. Loaded the Video component page in the bare-expo test app.
2. Confirmed that clicking "Open full screen" switched the video into fullscreen
  and logged `onFullscreenUpdate - {fullscreenUpdate: 1}` to the console.
3. Confirmed that clicking the native close fullscreen control switched the
  video out of fullscreen and logged
  `onFullscreenUpdate - {fullscreenUpdate: 3}` to the console.

Browsers:
* Chrome 89.0.4385.0 on Android 11 (emulated)
* Chrome 89.0.4389.90 on macOS
* Firefox 86.0.1 on macOS
* Safari 14.0.3 on macOS
* Safari on iOS 14.4 (simulator + iPhone SE)

For Safari (iOS and macOS) I manually executed the webkitExitFullScreen API from
console to confirm that it works as expected (there's no control in the test
app to trigger that API).

For IE11 I just used the dev console to manually test that the underlying DOM APIs/events that I used work as expected. I couldn't find an easy way to test my actual PR in IE11 because Expo doesn't work in that browser by default. I was
tempted to just drop the IE11 fullscreen code I came across entirely because a) Expo doesn't support IE11 out of the box and b) that IE11 code seemed to be broken anyway, but Chesterton's fence prevented me.

I did not write unit tests for the new WebFullscreenUtils module as that would
basically mean just testing the internal implementation details of the utilities
which is not very helpful.

I could not figure out how to get the `Video` e2e tests running for the web
application - I got strange errors when I tried that. As this is a cross browser
issue I'm not really sure running tests in puppeteer would tell us much anyway.
Ultimately something like BrowserStack is what's needed to test this change, but
that's beyond the scope of this PR.

fixes #12257

### CODE REVIEW

File: packages/expo-av/src/WebFullscreenUtils.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  if (supportsFullscreenAPI(element)) {
+    return element.requestFullscreen();
+  } else if (supportsWebkitFullscreenAPI(element)) {
+    element['webkitEnterFullScreen']?.();
+    return Promise.resolve();
+  } else if (supportsMsFullscreenAPI(element)) {
+    element['msRequestFullscreen']?.();
+    return Promise.resolve();
+  } else {
+    return Promise.reject(new Error('Fullscreen not supported'));
```
Comment (IjzerenHein) [inline packages/expo-av/src/WebFullscreenUtils.ts:44]: This is already an async function so no need for Promise.reject

```suggestion
    throw new Error('Fullscreen not supported');
```

File: packages/expo-av/src/WebFullscreenUtils.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Switches a video element into fullscreen.
+ */
+export async function requestFullscreen(element: HTMLMediaElement): Promise<void> {
+  if (supportsFullscreenAPI(element)) {
+    return element.requestFullscreen();
+  } else if (supportsWebkitFullscreenAPI(element)) {
+    element['webkitEnterFullScreen']?.();
+    return Promise.resolve();
```
Comment (IjzerenHein) [inline packages/expo-av/src/WebFullscreenUtils.ts:39]: `Promise.resolve()` can be removed here as `requestFullscreen` is an async function

File: packages/expo-av/src/WebFullscreenUtils.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  } else if (supportsMsFullscreenAPI(element)) {
+    element['msRequestFullscreen']?.();
+    return Promise.resolve();
```
Comment (IjzerenHein) [inline packages/expo-av/src/WebFullscreenUtils.ts:42]: `Promise.resolve()` can be removed here as `requestFullscreen` is an async function

File: packages/expo-av/src/WebFullscreenUtils.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  } else if (supportsWebkitFullscreenAPI(element)) {
+    element['webkitExitFullScreen']?.();
+    return Promise.resolve();
+  } else if (supportsMsFullscreenAPI(element)) {
+    document['msExitFullscreen']?.();
+    return Promise.resolve();
+  } else {
+    return Promise.reject(new Error('Fullscreen not supported'));
+  }
+}
```
Comment (IjzerenHein) [inline packages/expo-av/src/WebFullscreenUtils.ts:64]: No need to use Promise.resolve/reject here as well

File: packages/expo-av/src/WebFullscreenUtils.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  'requestFullscreen' in element;
+
+const supportsWebkitFullscreenAPI = (element: HTMLMediaElement): boolean =>
+  'webkitEnterFullScreen' in element;
+
+const supportsWebkitFullscreenChangeEvent = (): boolean => {
+  const element = document.createElement('video');
+  element.setAttribute('onwebkitfullscreenchange', 'return;');
+  return typeof element['onwebkitfullscreenchange'] === 'function';
+};
```
Comment (IjzerenHein) [inline packages/expo-av/src/WebFullscreenUtils.ts:11]: Please add comments or a code reference (stack overflow article?) on why this works and for which browsers

**Review by IjzerenHein — CHANGES_REQUESTED** [review]
Hi! This looks great! I verified locally and can confirm it works on all browsers now.

A couple things still need attention:
- Please remove any changes from the the root CHANGELOG.md and put them in `expo-av/CHANGELOG.md`
- Please rename the file `WebFullscreenUtils.ts` -> `FullscreenUtils.web.ts`

Awesome work! 👍

**Comment by elliotdickison** [comment]
* Moved changelog entry from CHANGELOG.md to packages/expo-av/CHANGELOG.md
* Renamed the file WebFullscreenUtils.ts to FullscreenUtils.web.ts
* Added additional documentation to the quirky feature detection code
* Removed all the redundant Promise.resolve/reject return statements

**Comment by elliotdickison** [comment]
@IjzerenHein Thanks for the feedback! All has been addressed. I think the blocked merge is due to the same GitHub action configuration issue that I experienced earlier.

File: packages/expo-av/CHANGELOG.md [hunk 6]
```diff
@@ -26,6 +26,8 @@ _This version does not introduce any user-facing changes._
 
 - Remove peerDependencies and unimodulePeerDependencies from Expo modules. ([#11980](https://github.com/expo/expo/pull/11980) by [@brentvatne](https://github.com/brentvatne))
 
+- Fixed the web Video Fullscreen APIs in Safari ([#12258](https://github.com/expo/expo/pull/12258) by [@elliotdickison](https://github.com/elliotdickison))
+
```
Comment (IjzerenHein) [inline packages/expo-av/CHANGELOG.md:30]: This entry is in the wrong place. It should be under `## Unpublished -> ### 🐛 Bug fixes`

**Review by IjzerenHein — CHANGES_REQUESTED** [review]
Awesome, thanks for the updates Eliot! There is one small thing, the changelog entry was added to an existing version but should be added to `Unpublished`.

File: packages/expo-av/CHANGELOG.md (hunk 6 above)
Comment (elliotdickison) [reply packages/expo-av/CHANGELOG.md]: doh! fixed

**Comment by Hirbod** [comment]
This PR will break the `<Video>` component on Web on UNMOUNT. See #14687

**Comment by Hirbod** [comment]
`addFullscreenListener()` in FullscreenUtils.web.js triggers on mount and unmount.
